### PR TITLE
Call graph based monomorphization

### PR DESCRIPTION
**Pull request by eliaslfox**
Current benchmarks computing the 20th fibonacci number on `Integer`:

```
Benchmarking COALTON-BENCHMARKS::RECURSIVE-FIB-MONOMORPHISED...
-                SAMPLES  TOTAL     MINIMUM   MAXIMUM   MEDIAN    AVERAGE   DEVIATION
REAL-TIME        1000     0.079599  0.000076  0.000261  0.000079  0.00008   0.00001
RUN-TIME         1000     0.079365  0.000076  0.000239  0.000079  0.000079  0.000009
USER-RUN-TIME    1000     0.078538  0.000075  0.000191  0.000078  0.000079  0.000008
SYSTEM-RUN-TIME  1000     0.001038  0         0.000026  0.000001  0.000001  0.000001
PAGE-FAULTS      1000     0         0         0         0         0         0.0
GC-RUN-TIME      1000     0         0         0         0         0         0.0
BYTES-CONSED     1000     16384     0         16384     0         16.384    517.84845
EVAL-CALLS       1000     0         0         0         0         0         0.0
done.
Benchmarking COALTON-BENCHMARKS::RECURSIVE-FIB-LISP...
-                SAMPLES  TOTAL     MINIMUM   MAXIMUM   MEDIAN    AVERAGE   DEVIATION
REAL-TIME        1000     0.069646  0.000066  0.000212  0.000067  0.00007   0.000014
RUN-TIME         1000     0.06955   0.000065  0.000197  0.000067  0.00007   0.000013
USER-RUN-TIME    1000     0.068565  0.000065  0.000197  0.000066  0.000069  0.000013
SYSTEM-RUN-TIME  1000     0.001048  0         0.000018  0.000001  0.000001  0.000001
PAGE-FAULTS      1000     0         0         0         0         0         0.0
GC-RUN-TIME      1000     0         0         0         0         0         0.0
BYTES-CONSED     1000     0         0         0         0         0         0.0
EVAL-CALLS       1000     0         0         0         0         0         0.0
done.
Benchmarking COALTON-BENCHMARKS::RECURSIVE-FIB-GENERIC...
-                SAMPLES  TOTAL     MINIMUM   MAXIMUM   MEDIAN    AVERAGE   DEVIATION
REAL-TIME        1000     0.398747  0.000383  0.000737  0.000393  0.000399  0.000027
RUN-TIME         1000     0.398435  0.000382  0.000731  0.000393  0.000398  0.000026
USER-RUN-TIME    1000     0.397163  0.000382  0.000703  0.000392  0.000397  0.000025
SYSTEM-RUN-TIME  1000     0.001323  0         0.000037  0.000001  0.000001  0.000002
PAGE-FAULTS      1000     0         0         0         0         0         0.0
GC-RUN-TIME      1000     0         0         0         0         0         0.0
BYTES-CONSED     1000     0         0         0         0         0         0.0
EVAL-CALLS       1000     0         0         0         0         0         0.0
done.
Benchmarking COALTON-BENCHMARKS::RECURSIVE-FIB...
-                SAMPLES  TOTAL     MINIMUM   MAXIMUM   MEDIAN    AVERAGE   DEVIATION
REAL-TIME        1000     0.081648  0.000077  0.000245  0.000077  0.000082  0.000019
RUN-TIME         1000     0.081557  0.000076  0.000247  0.000077  0.000082  0.000018
USER-RUN-TIME    1000     0.08045   0.000076  0.000241  0.000076  0.00008   0.000017
SYSTEM-RUN-TIME  1000     0.001112  0         0.000033  0.000001  0.000001  0.000002
PAGE-FAULTS      1000     0         0         0         0         0         0.0
GC-RUN-TIME      1000     0         0         0         0         0         0.0
BYTES-CONSED     1000     0         0         0         0         0         0.0
EVAL-CALLS       1000     0         0         0         0         0         0.0
done.
```
Current benchmarks computing the 10th fibonacci number on `Optional Integer`:

```
Benchmarking COALTON-BENCHMARKS::RECURSIVE-FIB-MONOMORPHISED-OPTIONAL...
-                SAMPLES  TOTAL     MINIMUM   MAXIMUM   MEDIAN    AVERAGE   DEVIATION
REAL-TIME        1000     0.022006  0.000015  0.000161  0.000018  0.000022  0.000007
RUN-TIME         1000     0.021876  0.000015  0.000136  0.000018  0.000022  0.000007
USER-RUN-TIME    1000     0.01747   0.000014  0.000124  0.000017  0.000017  0.000004
SYSTEM-RUN-TIME  1000     0.004438  0.000001  0.000025  0.000001  0.000004  0.000004
PAGE-FAULTS      1000     0         0         0         0         0         0.0
GC-RUN-TIME      1000     0         0         0         0         0         0.0
BYTES-CONSED     1000     40735200  32736     65472     32768     40735.2   8219.482
EVAL-CALLS       1000     0         0         0         0         0         0.0
done.
Benchmarking COALTON-BENCHMARKS::RECURSIVE-FIB-GENERIC-OPTIONAL...
-                SAMPLES  TOTAL     MINIMUM   MAXIMUM   MEDIAN    AVERAGE   DEVIATION
REAL-TIME        1000     0.026454  0.000021  0.000042  0.000026  0.000026  0.000002
RUN-TIME         1000     0.026425  0.000021  0.000042  0.000026  0.000026  0.000002
USER-RUN-TIME    1000     0.018423  0.000016  0.000033  0.000018  0.000018  0.000001
SYSTEM-RUN-TIME  1000     0.00795   0.000005  0.000019  0.000008  0.000008  0.000001
PAGE-FAULTS      1000     0         0         0         0         0         0.0
GC-RUN-TIME      1000     0         0         0         0         0         0.0
BYTES-CONSED     1000     40670032  32736     49152     32768     40670.03  8185.2974
EVAL-CALLS       1000     0         0         0         0         0         0.0
done.
```

Next steps:

- [x] Rerun type class resolution after applying substitutions. This will allow inlining more complex dictionaries.
- [x] Make method inlining compatible with dictionary hoisting.

### CODE REVIEW

**Comment by eliaslfox** [comment]
```
(coalton-codegen
  (define add (Num :a => :a -> :a -> :a))
  (define (add a b)
    (+ a b))

  (declare add-int (Integer -> Integer -> Integer))
  (define (add-int a b)
    (- a b))

  (specialize add add-int (Integer -> Integer -> Integer))

  (define x (add 1 2)))
```

File: src/coalton.lisp [hunk 1]
```diff
@@ -72,6 +78,10 @@ in FORMS that begin with that operator."
              (setf type (car type)))
 
            (setf (gethash type (getf plist 'repr-table)) (cons specifier arg)))
+
+         (handle-monomorphise (definition-name)
+           (push :monomorphise (gethash definition-name (getf plist 'attr-table))))
```
Comment (stylewarning) [inline src/coalton.lisp:83]: in general, quote symbols that are data

File: src/codegen/function-entry.lisp [hunk 2]
```diff
@@ -82,11 +82,13 @@
                                     (alexandria:format-symbol *package* "A~D" i))
                               funs)))
       `(progn
-         #+sbcl
-         (declaim (sb-ext:start-block ,@funs))
-         ,@(reverse body)
-         #+sbcl
-         (declaim (sb-ext:end-block))))))
+         (declaim (inline ,@funs))
+         (locally (declaim (notinline ,@funs))
```
Comment (stylewarning) [inline src/codegen/function-entry.lisp:86]: does this work? I think I've only seen
```
(declaim inline)
;; code
(declaim notinline)
```

do we really want these funcs to always be inlined after their definition?

**Review by stylewarning — APPROVED** [review]
monomorphize is probably preferred to monomorphise

File: src/codegen/function-entry.lisp (hunk 2 above)
Comment (eliaslfox) [reply src/codegen/function-entry.lisp]: It does not appear to actually work. I do think these functions, or at least the low numbered ones should be inlined. They are called on every indirect function call.

**Review by stylewarning — APPROVED** [review]

**Pull request merged**